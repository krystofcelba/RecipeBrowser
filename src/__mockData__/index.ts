export const base64Uri = 'data:image/jpeg;base64,xxxxxxxxx';

export const addRecipeFormValues = {
  name: 'recipe',
  description: 'desc',
  image: base64Uri,
  ingredients: [{ name: 'aa', amount: 1, unit: 'kg' }] as [{ name: string; amount: number; unit: string }],
  seasonings: ['red powder'] as [string],
  steps: ['ddd'] as [string],
};

export const mockedImagePickerImage = {
  path: '/xxx',
  mime: 'image/json',
  data: 'xxxxxxx',
};
