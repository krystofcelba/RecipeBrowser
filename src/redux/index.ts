import { applyMiddleware, createStore, compose, Store } from 'redux';
import { composeWithDevTools } from 'remote-redux-devtools';
import createSagaMiddleware from 'redux-saga';

import rootReducer, { RootState } from './reducers';
import rootSaga from './sagas';

const configureStore = (): Store<RootState> => {
  const sagaMiddleware = createSagaMiddleware();

  const composeEnhancers = __DEV__
    ? composeWithDevTools({
        realtime: true,
        hostname: 'localhost',
        port: 8000,
      })
    : compose;

  const store = createStore(
    rootReducer,
    composeEnhancers(applyMiddleware(sagaMiddleware)),
  );

  sagaMiddleware.run(rootSaga);
  return store;
};

export default configureStore;
