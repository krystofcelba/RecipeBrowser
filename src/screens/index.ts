import { Navigation } from 'react-native-navigation';

import RecipesList from './RecipesList';

export const registerScreens = (store, provider) => {
  Navigation.registerComponent(
    'recipeBrowser.RecipesList',
    () => RecipesList,
    store,
    provider,
  );
};
