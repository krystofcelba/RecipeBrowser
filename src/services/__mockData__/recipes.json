[
  {
    "description":
      "This buffet classic can also be healthy! Mayonnaise is replaced by cottage cheese. You can blend the filling for more creaminess.",
    "id": 9,
    "image": "https://nutrition-assets.freeletics.com/recipes/large/mustard_deviled_eggs.jpg",
    "ingredients": [
      {
        "amount": 4,
        "name": "hard-boiled eggs",
        "unit": ""
      },
      {
        "amount": 2,
        "name": "Dijon mustard",
        "unit": "tsp"
      },
      {
        "amount": 100,
        "name": "cottage cheese",
        "unit": "g"
      },
      {
        "amount": 2,
        "name": "chopped dill",
        "unit": "tbsp"
      }
    ],
    "name": "Mustard Deviled Eggs",
    "seasonings": ["Paprika powder"],
    "steps": [
      "Halve eggs and remove yolk.",
      "Mix yolks with mustard, cottage cheese and half of dill.",
      "Fill egg halves with yolk mixture and garnish with remaining dill."
    ]
  }
]
