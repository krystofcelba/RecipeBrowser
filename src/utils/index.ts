import i18n from 'src/assets/i18n';
import { Map } from 'immutable';

export const isString = value => typeof value === 'string';
export const isNumber = value => !isNaN(Number(value));

export const formatImageBase64Uri = ({ mime, data }: { mime: string; data: string }) => `data:${mime};base64,${data}`;
export const extractMimeType = (base64Uri: string) => base64Uri.slice(5, base64Uri.indexOf(';base64,'));

export const validateNewRecipeFormValues = (values: Map<string, any>) => {
  let errors = {};
  const { name, description, ingredients, seasonings, steps } = values.toJS();
  if (!name) {
    errors = { ...errors, name: i18n.t('requiredError', { field: i18n.t('name') }) };
  }
  if (!description) {
    errors = { ...errors, description: i18n.t('requiredError', { field: i18n.t('description') }) };
  }

  if (!ingredients) {
    errors = {
      ...errors,
      ingredients: { _error: i18n.t('atLeastOneError', { field: i18n.t('ingredient').toLowerCase() }) },
    };
  } else {
    const ingredientsErrors = ingredients.map(value => {
      const { name, amount, unit } = value;
      let err = '';
      if (!name) {
        err = `${i18n.t('requiredError', { field: i18n.t('name') })}. `;
      }
      if (!amount) {
        err = `${err}${i18n.t('requiredError', { field: i18n.t('amount') })}. `;
      } else if (amount && !isNumber(amount)) {
        err = `${err}${i18n.t('mustBeNumberError', { field: i18n.t('amount') })}. `;
      }
      if (!unit) {
        err = `${err}${i18n.t('requiredError', { field: i18n.t('unit') })}. `;
      }
      return err;
    });
    errors = { ...errors, ingredients: ingredientsErrors };
  }

  if (!seasonings) {
    errors = {
      ...errors,
      seasonings: { _error: i18n.t('atLeastOneError', { field: i18n.t('seasoning').toLowerCase() }) },
    };
  } else {
    const arrayErrors = seasonings.map(value => {
      return !isString(value) ? `${i18n.t('requiredError', { field: i18n.t('value') })}. ` : '';
    });
    errors = { ...errors, seasonings: arrayErrors };
  }

  if (!steps || !steps.length) {
    errors = {
      ...errors,
      steps: { _error: i18n.t('atLeastOneError', { field: i18n.t('step').toLowerCase() }) },
    };
  } else {
    const arrayErrors = steps.map(value => {
      return !isString(value) ? `${i18n.t('requiredError', { field: i18n.t('value') })}. ` : '';
    });
    errors = { ...errors, steps: arrayErrors };
  }

  return errors;
};
