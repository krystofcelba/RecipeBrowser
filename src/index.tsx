import { Component } from 'react';
import { Provider } from 'react-redux';
import { Navigation } from 'react-native-navigation';

import configureStore from './redux';
import { registerScreens } from './screens';
import { navigatorStyle } from './assets/constants';

const store = configureStore();
registerScreens(store, Provider);

class App extends Component {
  constructor(props) {
    super(props);
    this.startApp();
  }

  startApp = () => {
    Navigation.startSingleScreenApp({
      screen: {
        screen: 'recipeBrowser.RecipesList',
        ...navigatorStyle,
      },
      appStyle: {
        keepStyleAcrossPush: false,
      },
    });
  };
}

export default App;
