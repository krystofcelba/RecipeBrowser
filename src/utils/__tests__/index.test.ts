import { Map } from 'immutable';

import { base64Uri, addRecipeFormValues } from 'src/__mockData__';
import { isString, isNumber, extractMimeType, validateNewRecipeFormValues } from '../';

describe('utils functions', () => {
  it('should return true if it is string, false otherwise', () => {
    expect(isString('test')).toBe(true);
    expect(isString(Object('test'))).toBe(false);
  });

  it('should return true if it is number, false otherwise', () => {
    expect(isNumber(9)).toBe(true);
    expect(isNumber('10')).toBe(true);
    expect(isNumber('10.1')).toBe(true);
    expect(isNumber('10.a')).toBe(false);
  });

  it('should extract and return mime type from base 64 data uri', () => {
    expect(extractMimeType(base64Uri)).toBe('image/jpeg');
  });
});

describe('validateNewRecipeFormValues', () => {
  it('should return validations errors as expected', () => {
    expect(validateNewRecipeFormValues(Map({}))).toMatchSnapshot();

    expect(validateNewRecipeFormValues(Map({ ingredients: [] }))).toMatchSnapshot();
    expect(validateNewRecipeFormValues(Map({ ingredients: [{}] }))).toMatchSnapshot();
    expect(validateNewRecipeFormValues(Map({ ingredients: [{ amount: 'a' }] }))).toMatchSnapshot();

    expect(validateNewRecipeFormValues(Map({ seasonings: [''], steps: [''] }))).toMatchSnapshot();
    expect(validateNewRecipeFormValues(Map({ seasonings: [Object('')], steps: [Object('')] }))).toMatchSnapshot();

    expect(validateNewRecipeFormValues(Map(addRecipeFormValues))).toMatchSnapshot();
  });
});
