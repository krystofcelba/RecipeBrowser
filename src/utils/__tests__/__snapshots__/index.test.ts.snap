// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateNewRecipeFormValues should return validations errors as expected 1`] = `
Object {
  "description": "Description is required",
  "ingredients": Object {
    "_error": "At least one ingredient must be entered",
  },
  "name": "Name is required",
  "seasonings": Object {
    "_error": "At least one seasoning must be entered",
  },
  "steps": Object {
    "_error": "At least one step must be entered",
  },
}
`;

exports[`validateNewRecipeFormValues should return validations errors as expected 2`] = `
Object {
  "description": "Description is required",
  "ingredients": Array [],
  "name": "Name is required",
  "seasonings": Object {
    "_error": "At least one seasoning must be entered",
  },
  "steps": Object {
    "_error": "At least one step must be entered",
  },
}
`;

exports[`validateNewRecipeFormValues should return validations errors as expected 3`] = `
Object {
  "description": "Description is required",
  "ingredients": Array [
    "Name is required. Amount is required. Unit is required. ",
  ],
  "name": "Name is required",
  "seasonings": Object {
    "_error": "At least one seasoning must be entered",
  },
  "steps": Object {
    "_error": "At least one step must be entered",
  },
}
`;

exports[`validateNewRecipeFormValues should return validations errors as expected 4`] = `
Object {
  "description": "Description is required",
  "ingredients": Array [
    "Name is required. Amount must be a number. Unit is required. ",
  ],
  "name": "Name is required",
  "seasonings": Object {
    "_error": "At least one seasoning must be entered",
  },
  "steps": Object {
    "_error": "At least one step must be entered",
  },
}
`;

exports[`validateNewRecipeFormValues should return validations errors as expected 5`] = `
Object {
  "description": "Description is required",
  "ingredients": Object {
    "_error": "At least one ingredient must be entered",
  },
  "name": "Name is required",
  "seasonings": Array [
    "",
  ],
  "steps": Array [
    "",
  ],
}
`;

exports[`validateNewRecipeFormValues should return validations errors as expected 6`] = `
Object {
  "description": "Description is required",
  "ingredients": Object {
    "_error": "At least one ingredient must be entered",
  },
  "name": "Name is required",
  "seasonings": Array [
    "Value is required. ",
  ],
  "steps": Array [
    "Value is required. ",
  ],
}
`;

exports[`validateNewRecipeFormValues should return validations errors as expected 7`] = `
Object {
  "ingredients": Array [
    "",
  ],
  "seasonings": Array [
    "",
  ],
  "steps": Array [
    "",
  ],
}
`;
